api/admin/controller.py

```diff
@@ -179,33 +181,42 @@ def admin_auth_provider(self, type):
                 return provider
         return None
 
-    def authenticated_admin_from_request(self, email=None, type=None):
+    def authenticated_admin_from_request(self):
         """Returns an authenticated admin or a problem detail."""
         if not self.admin_auth_providers:
             return ADMIN_AUTH_NOT_CONFIGURED
 
-        email = email or flask.session.get("admin_email")
-        type = type or flask.session.get("auth_type")
+        email = flask.session.get("admin_email")
+        type = flask.session.get("auth_type")
 
         if email and type:
             admin = get_one(self._db, Admin, email=email)
             auth = self.admin_auth_provider(type)
             if not auth:
                 return ADMIN_AUTH_MECHANISM_NOT_CONFIGURED
             if admin and auth.active_credentials(admin):
+                flask.request.admin = admin
                 return admin
+        flask.request.admin = None
         return INVALID_ADMIN_CREDENTIALS
 
     def authenticated_admin(self, admin_details):
         """Creates or updates an admin with the given details"""
 
-        admin, ignore = get_one_or_create(
+        admin, is_new = get_one_or_create(
             self._db, Admin, email=admin_details['email']
         )
         admin.update_credentials(
             self._db,
             credential=admin_details.get('credentials'),
         )
+        if is_new and admin_details.get("roles"):
+            for role in admin_details.get("roles"):
+                if role.get("role") in AdminRole.ROLES:
+                    library = Library.lookup(self._db, role.get("library"))
+                    admin.add_role(role.get("role"), library)
```

Comment:
My previous comment is outdated -- this code must have moved. This is the bit I'm talking about. Elsewhere in the code you can see the transaction rolled back if the library lookup fails.